### PR TITLE
CORE-4356 CPB Tool

### PR DESCRIPTION
**Pull request by kyriathar**
- packs specified Cpks into a Cpb.
- adds and populates manifest file with specified attributes.
- signs Cpb. 

Please note that [CORE-4356](https://r3-cev.atlassian.net/browse/CORE-4356) also describes extra signing capabilities but these should be a different packaging subcommand and therefore will be addressed under a different ticket ([CORE-5741](https://r3-cev.atlassian.net/browse/CORE-5741)) as it is not part of this flow/ subcommand.

### CODE REVIEW

File: plugins/package/cpb/gradle.properties
Comment (pmck91) [inline plugins/package/cpb/gradle.properties]: please change to R3 Ltd.

**Comment by driessamyn** [comment]
What's the reason for this to be in this repo?
Is there guidance somewhere for when we put plugins in `corda-runtime-os` and when not?

We should not have plugins in this repo IMO, either the repo name is wrong, or plugins must be somewhere else.

File: plugins/package/cpb/gradle.properties
Comment (kyriathar) [reply plugins/package/cpb/gradle.properties]: This is as above ("pluginProvider=R3") in all plugins. I guess we should be changing all of them?

Comment (driessamyn) [reply plugins/package/cpb/gradle.properties]: Let's leave that out-of-scope for this PR. Consistency is good and if we want to change the rule, let's do that in a separate PR so the change is clear.

**Comment by josephzunigadaly** [comment]
> What's the reason for this to be in this repo? Is there guidance somewhere for when we put plugins in `corda-runtime-os` and when not?
> 
> We should not have plugins in this repo IMO, either the repo name is wrong, or plugins must be somewhere else.

I added this to the repo. I've since spoken to Ben about calling libraries in runtime-os and he says this needs to move into runtime-os to access those libraries. We have a plugins folder over there already. So we do need to move this.

I found this after Kyriakos had started work so my plan was to let this complete then do the move before starting on the next plugin task.

**Comment by kyriathar** [comment]
This is a good point. I think we said with @josephzunigadaly that we should be moving them (at least CPx ones but probably all of them) in `corda-runtime-os`. But perhaps in a subsequent PR. @josephzunigadaly could you please confirm.

File: plugins/package/cpb/src/main/kotlin/net/corda/cli/plugins/packaging/cpb/CreateCpbPluginWrapper.kt
Comment (kyriathar) [inline plugins/package/cpb/src/main/kotlin/net/corda/cli/plugins/packaging/cpb/CreateCpbPluginWrapper.kt]: As discussed offline with @josephzunigadaly create Cpb command should be made a sub command under existing packaging plugin.

File: plugins/package/README.md
Comment (josephzunigadaly) [inline plugins/package/README.md]: I think we should save this rename for a later date. When we do this, we need to tell the whole development team. They are actively using this.

We also need to move this whole plugin into corda-runtime-os so I think it's best to combine the changes so we only communicate the changes once. Otherwise we'll give the impression that this is changing often and is unreliable.

File: plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt
Comment (josephzunigadaly) [inline plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt]: In the help text, we should use all caps "CPKs", "CPB"

File: plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt
Comment (josephzunigadaly) [inline plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt]: We could do this check on all CPKs before starting work

File: plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt
Comment (josephzunigadaly) [inline plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt]: We should delete this file once we are done. `buildAndSignCpiJar` has one approach to this but it won't fit with the structure we have here.

File: plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt
Comment (josephzunigadaly) [inline plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt]: We are mixing the Zip API with the Jar API. There is some inheritance between the APIs, so it works, but this should be a `JarEntry`.

File: plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/signing/SigningHelpers.kt
Comment (josephzunigadaly) [inline plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/signing/SigningHelpers.kt]: Can these live inside `CpxSigner` instead of a different `SigningHelpers` object?

It seems these are related to the work `CpxSigner` does.

File: plugins/package/src/test/kotlin/net/corda/cli/plugins/packaging/CreateCpbTest.kt
Comment (josephzunigadaly) [inline plugins/package/src/test/kotlin/net/corda/cli/plugins/packaging/CreateCpbTest.kt]: This function can be moved outside of the companion object. It is private, always accessed from an instance of the class and does not use any static state.

File: settings.gradle
Comment (josephzunigadaly) [inline settings.gradle]: Does this add BLT as code owners? If so, lets revert this so there are no changes to gradle files in the PR

**Comment by pmck91** [comment]
> What's the reason for this to be in this repo? Is there guidance somewhere for when we put plugins in `corda-runtime-os` and when not?
> 
> We should not have plugins in this repo IMO, either the repo name is wrong, or plugins must be somewhere else.

I agree, @kyriathar I would ask that you move this into runtime os as well. Moving forward we will more than likely always use runtime os for the plugins, unless someone develops a stand alone one for some (un-corda) reason. 

If you have any questions about how the build.gradle will look for the runtime os plugins don't be afraid to ask, as it uses a gradle plugin to bundle everything.

File: plugins/package/README.md
Comment (kyriathar) [reply plugins/package/README.md]: Good catch, will revert this change.

File: plugins/package/build.gradle
Comment (josephzunigadaly) [inline plugins/package/build.gradle]: Can we do the job without this library?

File: plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt
Comment (kyriathar) [reply plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt]: Done

File: plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt
Comment (kyriathar) [reply plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt]: Agreed, in which case `Sequence` does not make much sense anymore, will change to a list.

Comment (kyriathar) [reply plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt]: Done

File: plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt
Comment (kyriathar) [reply plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt]: Done

File: plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt
Comment (kyriathar) [reply plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt]: Changed to `JarEntry`.

File: plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/signing/SigningHelpers.kt
Comment (kyriathar) [reply plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/signing/SigningHelpers.kt]: Done

File: plugins/package/src/test/kotlin/net/corda/cli/plugins/packaging/CreateCpbTest.kt
Comment (kyriathar) [reply plugins/package/src/test/kotlin/net/corda/cli/plugins/packaging/CreateCpbTest.kt]: Moved it to be a static function in the test class because it does not make any use of test class state either.

File: settings.gradle
Comment (kyriathar) [reply settings.gradle]: Done

File: plugins/package/build.gradle
Comment (kyriathar) [reply plugins/package/build.gradle]: I 'd say it would be useful to have because of its assertion options on expected/ actual sets. Unless there is such option in existing test libraries that I have missed.

File: plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt
Comment (josephzunigadaly) [inline plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt]: The answer to the question is no, it's not required. It should default to `false`.

Can we change the description to "Allow upgrade without flow draining" please?

File: plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt
Comment (josephzunigadaly) [inline plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt]: We don't need to mention "Manifest attribute" in the description. It describes how the implementation will store the values which is an implementation detail. "CPB name"/"CPB version" would be enough.

File: plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt
Comment (josephzunigadaly) [inline plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt]: We don't need the random UUID here. `createTempFile` will give us an unused filename.

File: plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt
Comment (kyriathar) [reply plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt]: Done

File: plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt
Comment (kyriathar) [reply plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt]: Good point, will address.

File: plugins/package/README.md
Comment (josephzunigadaly) [inline plugins/package/README.md]: Does it work without the quotes? We want tab completion to help users enter the filenames. We might need to look at how the parameter is configured if we need the quotes.

Comment (kyriathar) [reply plugins/package/README.md]: It did not work without the quotes. Good point thanks! Let me see if I can make it work without the quotes.

File: plugins/package/src/test/kotlin/net/corda/cli/plugins/packaging/CreateCpbTest.kt
Comment (josephzunigadaly) [inline plugins/package/src/test/kotlin/net/corda/cli/plugins/packaging/CreateCpbTest.kt]: Can we pass the filename into the method as a parameter?

File: plugins/package/src/test/kotlin/net/corda/cli/plugins/packaging/CreateCpbTest.kt
Comment (josephzunigadaly) [inline plugins/package/src/test/kotlin/net/corda/cli/plugins/packaging/CreateCpbTest.kt]: `tempDir` will be a new empty directory for every test. We could call the output file `cpb-test-outcome.cpb` because it will not see the files from other tests and other tests will not see it.

> _This only works when `tempDir` is on a non-static variable, which is the case here_

File: plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt
Comment (josephzunigadaly) [inline plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt]: Does changing the type to boolean work?

Comment (kyriathar) [reply plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt]: Done

File: plugins/package/README.md
Comment (kyriathar) [reply plugins/package/README.md]: Done -passed in cpks are made `Parameters` to allow auto-completion-.

File: plugins/package/README.md
Comment (vkolomeyko) [inline plugins/package/README.md:16]: Can we please move all this functionality to `corda-runtime-os` repo under: https://github.com/corda/corda-runtime-os/tree/release/os/5.0/tools/plugins ?

**Review by vkolomeyko — DISMISSED** [review]

File: plugins/package/README.md
Comment (kyriathar) [reply plugins/package/README.md:16]: Once this is merged, next task is to move the package plugin to `corda-runtime-os` repo.

**Review by josephzunigadaly — APPROVED** [review]

File: plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt
Comment (barrylapthorn) [inline plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt:107]: Is this meant to be hard-coded?  Or something that @josephzunigadaly knows about?

Comment (kyriathar) [reply plugins/package/src/main/kotlin/net/corda/cli/plugins/packaging/CreateCpb.kt:107]: Thanks for the catch, since this is now merged will address in subsequent PR.

**Pull request merged**
